### PR TITLE
[compression] pruning dependency v2

### PR DESCRIPTION
**Pull request by super-dainiu**
#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

**Comment by J-shang** [comment]
please design a representative model (should have all dependencies we try to deal with) and test dependency on it, add this into unit test.

**Comment by super-dainiu** [comment]
I see

File: nni/compression/pytorch/speedup/v2/dependency.py [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+    for key, value in dependency.items():
+        if key not in visited:
+            tmp = set([key]) | value
+            visited.update(tmp)
+            if len(tmp) > 1:
+                d_sets.append(tmp)
+    return d_sets
+
+
+def build_weight_sharing_dependency(graph_module: torch.fx.GraphModule) -> List[List[torch.fx.Node]]:
```
Comment (zheng-ningxin) [inline nni/compression/pytorch/speedup/v2/dependency.py:230]: Could you please explain more about the weight-sharing dependency? What scenarios is it for? why it has such dependencies?

File: nni/compression/pytorch/speedup/v2/dependency.py [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+            return meta[0][0]
+        return None
+
+    in_shape = get_shape(find_parent(node))
+    out_shape = get_shape(node)
+    if not in_shape or not out_shape or len(in_shape) <= 1 or len(out_shape) <= 1:
+        return True
+    in_channel = in_shape[1]
+    out_channel = out_shape[1]
+    return in_channel != out_channel
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/v2/dependency.py:105]: Is it possible that the number of channels is the same but it is a break operation?

File: nni/compression/pytorch/speedup/v2/dependency.py [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+
+            elif node.op == 'call_function':
+                if node.target in CALL_FUNCTION_REDUCE:
+                        # refer issue 4540 for more details. Multiplication actually
+                        # will not introduce the channel dependency, cause the misaligned
+                        # channels can propagate to each other. However, when one of the input
+                        # tensor is from skip connection(residual), the channel propagation
+                        # may be failed(the input is also used by another layer and cannot be
+                        # pruned), in this case, we need to fix the conflict maunally.
+                        d_set = set(find_adjacent_layers(node, graph_module, target_types, 'parent'))
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/v2/dependency.py:331]: indentation

File: nni/compression/pytorch/speedup/v2/dependency.py (hunk 1 above)
Comment (super-dainiu) [reply nni/compression/pytorch/speedup/v2/dependency.py:230]: this sounds like
```python
class Example(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.conv = nn.Conv2d(3, 3, 3, padding=1)
        self.conv2 = nn.Conv2d(3, 3, 3, padding=1)
        
    def forward(self, x):
        x = self.conv(x) + x
        x = self.conv2(x) + x
        x = self.conv(x) + x
        return x
```
where `self.conv` is repeatedly used in forward. in `torch.fx` we might prune ``self.conv`` twice.

File: nni/compression/pytorch/speedup/v2/dependency.py (hunk 2 above)
Comment (super-dainiu) [reply nni/compression/pytorch/speedup/v2/dependency.py:105]: it's really hard for me to find an example. maybe I add a TODO here?

**Review by J-shang — APPROVED** [review]

**Review by zheng-ningxin — APPROVED** [review]
LGTM

**Pull request merged**